feat(writer): add gif animation writer

MarwinNumbers opened the pull request:
Method for api to easy generate gif animations out of rendered frames.

closes iss #481

----------------------------------------------------------------------------------------------
Hey, I hope my suggestion is useful and fits your requirements. 

I have written an extension to easily generate a gif animation out of your rendered data. It can be applied intuitively in the same way the `bproc.writer.write_hdf5()` method works. I have also tried several settings like stereo rendering, normals, depth, segmentation, and it all worked just fine. Additionally I made a new example in `examples/advanced/gif_animation` to show not just how to get a gif file, but also how to get similar settings like in the home page animation from BlenderProc, because I think it is a good insight for new people using this software. At least for me it was a big help to get quick and easy animations when I run through all the possibilities BlenderProc provides. Not done so far is the .yaml file in the example and overall the compatibility to it.  Thats because I just used the BlenderProc python API. To show results I have here some further animations, which I didn't plug into the image folder to keep it slim. 

![99f_colors_animation](https://user-images.githubusercontent.com/97362660/153773388-35ae7ceb-71d1-41e5-92ff-48a207b1240c.gif)
![99f_depth_animation](https://user-images.githubusercontent.com/97362660/153773398-c36835ac-63b8-4d6b-b027-9fbff1161785.gif)
![99f_instance_segmaps_animation](https://user-images.githubusercontent.com/97362660/153773407-0b9ef667-1e44-4324-966f-810e08cf51a2.gif)
![99f_normals_animation](https://user-images.githubusercontent.com/97362660/153773416-aedaf4b7-c67f-47c1-9d0d-9829654329ac.gif)
![1_forward_flow_animation](https://user-images.githubusercontent.com/97362660/153773448-c5e1c003-096f-423d-9a1d-7bc569ee8644.gif)

The last picture shows the optical forward flow of the scene. 

I chose the `python/writer/GifWriter.py` location to implement it, since it seemed the most natural choice. I tried to stay similar to the other `#Writer.py` modules and code in naming and ordering variables. I also took a look at the recommended PEP8 guideline, why the code seems to be in my point of view kind of compressed since the "only 79 for each line" rule :)

As storage location I chose a separate folder `output/gif_animations` and the naming is like `<number>_<render-key>_<left or right if stereo>_animation.gif`.

# How the code works:
* Create an output-folder
* Create temporary .png collections in temporary folder
* Merge the collections to a .gif file each
* Delete temporary data

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+import matplotlib.pyplot as plt
+
+try:
+    from visHdf5Files import vis_data
```
This nearly always throws an exception. Why even try to import it this way?

MarwinNumbers replied:
Right, I have changed it.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+        output_dir_path: str,
+        output_data_dict: Dict[str, List[Union[np.ndarray, list, dict]]],
+        append_to_existing_output: bool = True,
+        time_constant: int = 50,
```
Unlucky parameter name, `animation_speed` or `animation_visibility_in_ms`? The parameter name should tell you roughly what it is without consulting the documentation.

MarwinNumbers replied:
Now its called `frame_duration_in_ms` :)

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+        output_data_dict: Dict[str, List[Union[np.ndarray, list, dict]]],
+        append_to_existing_output: bool = True,
+        time_constant: int = 50,
+        time_direction: bool = False):
```
Same here, `reverse_gif` tells you directly what it does.

MarwinNumbers replied:
Now its called `reverse_animation`

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+        return output_dir_path
+
+    @staticmethod
+    def _select_keys(
+            output_data_dict: Dict[str, List[Union[np.ndarray, list, dict]]]
+            ) -> List[str]:
+        """ Sorts out keys which are just metadata and not plottable """
+        keys_to_use = []
+        for key in output_data_dict.keys():
+            value = output_data_dict[key][0]
+            value = np.array(value)
+            if np.issubdtype(value.dtype, np.string_) or len(value.shape) == 1:
+                pass # metadata
+            else:
+                keys_to_use.append(key)
+        return keys_to_use
```
```suggestion
    def _select_keys(
            output_data_dict: Dict[str, List[Union[np.ndarray, list, dict]]]
            ) -> List[str]:
        """ Sorts out keys which are just metadata and not plottable """
        def is_image(x: Union[np.ndarray, list, dict]) -> bool:
           """ Checks if the input x is not a string and is not a vector"""
            x = np.array(x)
            return not np.issubdtype(value.dtype, np.string_) and len(value.shape) != 1
       return [key for key, value in output_data_dict.items() if len(value) > 0 and is_image(value[0])]
```

This code is quite long and convoluted. You should not have empty if passes just negate the if statement then. Accessing the first element of a list without checking its length is also dangerous.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+                    for number, frame in enumerate(output_data_dict[key]):
+                        vis_data(key=key,
+                                data=frame[index], 
+                                save_to_file=tmp_dir.name + f'/{number}.png')
```
```suggestion
                                save_to_file=os.path.join(tmp_dir.name, f'{number}.png'))
```

MarwinNumbers replied:
I have done it like you wrote it down.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+                for number, frame in enumerate(output_data_dict[key]):
+                    vis_data(key=key, 
+                            data=frame, 
+                            save_to_file=tmp_dir.name + f'/{number}.png')
```
Same here. No merging of paths without a library (Windows)

MarwinNumbers replied:
Done. Like you said.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+                            + perspective 
+                            + ': ' 
+                            + tmp_dir.name)
+                    to_animate.update({key + perspective: tmp_dir})
```
```suggestion
                    to_animate[key + perspective] = tmp_dir
```
This is the preferred way of updating elements in a dict. Is also much faster as no new dict has to be generated each time.

MarwinNumbers replied:
Done. Like you said.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+            if value.shape[0] is 2:
+                # stereo images
+                for index, perspective in enumerate(['_L', '_R']):
+                    tmp_dir = tempfile.TemporaryDirectory()
```
Please use the BlenderProc temp directory: https://github.com/DLR-RM/BlenderProc/blob/550faebfea559ae4bbca893df779aad2501062ce/blenderproc/python/utility/Utility.py#L174

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+        """
+
+        if not set(keys_to_use) <= set(output_data_dict.keys()):
+            raise ValueError("The keys_to_use list must be contained in"
+                            "the list of keys from the output_data_dict!")
```
The error message contains more info than actually is being checked. There is no check if there elements which are not in the output keys. 

```suggestion
            raise ValueError("There are more requested keys than are available in the output data.")
```

MarwinNumbers replied:
It's actually true what I wrote. It checks if it is a subset. So it's like:
```python
set([1,4]) <= set([1,2,3])
```
```
>>> False
```
It does not behave like a `len(list)` comparison.

themasterlink replied:
Interesting I haven't seen that before.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+        """
+        for key, tmp_dir in to_animate.items():
+            print('gif for ' + key + ' at ' + tmp_dir.name)
+            frames = []
+            # Loads all .png images as paths to a list
+            frame_paths = glob.glob(tmp_dir.name + '/*.png')
+            # sorts list in size w.r.t. given key
+            frame_paths.sort(key = GifWriterUtility._sort_method, 
+                            reverse=time_direction)
```
Why are they loaded again, did we not just save them. Why did not we not save them in a list, then now we wouldn't have to sort them again.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+    @staticmethod
+    def _sort_method(item) -> int:
+        """ Order the images due to the file numbering """
+        _, file_name = item.rsplit('/', maxsplit=1)
```
This does not work on Windows.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+        return to_animate
+
+    @staticmethod
+    def _sort_method(item) -> int:
+        """ Order the images due to the file numbering """
+        _, file_name = item.rsplit('/', maxsplit=1)
+        file_number, _ = file_name.split('.')
+        return int(file_number)
```
This method does not sort anything. It only extracts the number of the file.

At the top: `from pathlib import Path`

```suggestion
    def _extract_number(path: str) -> int:
        """ Order the images due to the file numbering """
        value = Path(path).with_suffix("").name
        if value.isnumeric():
            return int(value)
        else:
            raise Exception(f"The input item is not convertible to an int: {item}")
```

MarwinNumbers replied:
I thought it would be just fine, since I always started at the path ending with `rsplit()` instead of `split`. But I will handle it like you mentioned in the future :)

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+                frames.append(new_frame)
+
+            name_ending = f"_{key}_animation.gif"
+            gif_number = GifWriterUtility._look_for_existing_output(
+                                            output_dir_path, 
+                                            append_to_existing_output, 
+                                            name_ending)
```
Why again can we not save this information? Why do we need to check the file system if the file was already created?

MarwinNumbers replied:
Oh, this here is a different story. At one side, we work with temporary .png files created through matplotlib and here we are working with .gif files in a different folder, the  output folder. This line and the `_look_for_exisiting_output()` method are well behaving, since it checks if there was a previous run in the output folder, such that the user has no chance to delete any produced .gif files by mistake.

themasterlink replied:
BlenderProc does not take care of existing data, if the user specifies the same folder again the existing data will be overwritten.

themasterlink commented on blenderproc/python/writer/GifWriterUtility.py:
```diff
@@ -0,0 +1,210 @@
+                                            name_ending)
+            file_name = str(gif_number) + name_ending
+            file = os.path.join(output_dir_path, file_name)
+            frames[0].save(file,
+                            format ='GIF',
+                            append_images = frames[1:],
+                            save_all = True,
+                            duration = time_constant,
+                            loop = 0)
```
Does this use the first loaded png image? Why does the final gif contain the `gif_number`?

themasterlink reviewed (changes requested):
I like the concept, but I am confused about the execution. Often times it is much more convoluted than necessary. Which makes it harder to read and harder to repair/change in the future.

In my head I don't see a reason, why the images shouldn't be directly written into the gif file format. It is weird that you first write them to png and then they are loaded again and written into a gif?

MarwinNumbers commented:
@themasterlink About the question, why saving .png in-between. This is because I decided to use the `visHdf5Files.py` module, which does the formatting with matplotlib of all pictures consistent in BlenderProc. So my motivation was:
1. If you change the style of images in `visHdf5Files.py`, you do not have to care about the `GifWriterUility.py` module separately. 
2. At least on my machine, the time consumption for executing the `write_gif_animation()` was negligible to the render-time.
So I kept the way this already working system. I am not sure if this reasoning works fine for you too :)

MarwinNumbers commented:
@themasterlink 
> So, what I would change, I would save the paths to all generated png files, I would store them in the BlenderProc temp folder that way they get removed even if BlenderProc crashes. Then I would load them in again and save them into a gif file. Without checking if the gif file was created before.

Your idea worked just great. I have already implemented it. Now the `glob` module is gone, the `tempfile` module is gone, we got rid off some lines of code and the internal bproc temp-directory is used. 

> Without checking if the gif file was created before.

Well, there is a point. I don't want to bother you with that, but the `WriterUtility.py` and therefore the `write_Hdf5()` method do have the parameter `append_to_existing_output: bool = False` which just does the same. I implemented it because of symmetry. And if people want to loop over different settings, like drop different things on the head of the monkey, the `append_to_existing_output: bool` parameter would be useful. Just my thoughts...

themasterlink commented on examples/advanced/gif_animation/main.py:
```diff
@@ -0,0 +1,99 @@
+light2.set_location([-5,-5,5])
+
+# Define the rythm of the blinking
+frame = 0
+on_frames = []
+while frame < frame_end:
+    on_frames.append(frame)
+    frame += 1
+    if frame % 10 == 0:
+        frame += 10
+
+for frame in range(frame_end):
+    if frame in on_frames:
+        light2.set_energy(600, frame=frame)
+    else:
+        light2.set_energy(0, frame=frame)
```
It took me a moment to get what it is happening here, but you are trying to turn the light on for ten frames and then turn it off for ten frames?

```suggestion
for frame in range(frame_end):
    # this changes every ten frames from on to off and vice-versa
    if (frame // 10) % 2 == 0:
        light2.set_energy(600, frame=frame)
    else:
        light2.set_energy(0, frame=frame)
```

themasterlink reviewed (changes requested):
One minor thing left
